common/meta/types/src/user_privilege.rs

```diff
@@ -25,35 +26,105 @@ use enumflags2::BitFlags;
 pub enum UserPrivilegeType {
     // UsagePrivilege is a synonym for “no privileges”
     Usage = 1 << 0,
-    // Privilege to create databases and tables.
-    Create = 1 << 1,
     // Privilege to select rows from tables in a database.
     Select = 1 << 2,
     // Privilege to insert into tables in a database.
     Insert = 1 << 3,
-    // Privilege to SET variables.
+    // Privilege to update rows in a table
+    Update = 1 << 5,
+    // Privilege to delete rows in a table
+    Delete = 1 << 6,
+    // Privilege to create databases or tables.
+    Create = 1 << 1,
+    // Privilege to drop databases or tables.
+    Drop = 1 << 7,
+    // Privilege to delete rows in a table
+    Alter = 1 << 8,
+    // Privilege to Kill query, Set global configs, etc.
+    Super = 1 << 9,
+    // Privilege to Create User.
+    CreateUser = 1 << 10,
+    // Privilege to Create Role.
+    CreateRole = 1 << 11,
+    // Privilege to Grant/Revoke privileges to users or roles
+    Grant = 1 << 12,
+    // TODO: remove this later
     Set = 1 << 4,
 }
 
 const ALL_PRIVILEGES: BitFlags<UserPrivilegeType> = make_bitflags!(
-    UserPrivilegeType::{Create
+    UserPrivilegeType::{
+        Create
         | Select
         | Insert
-        | Set}
+        | Update
+        | Delete
+        | Drop
+        | Alter
+        | Super
+        | CreateUser
+        | CreateRole
+        | Grant
+    }
 );
 
 impl std::fmt::Display for UserPrivilegeType {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
         write!(f, "{}", match self {
             UserPrivilegeType::Usage => "USAGE",
             UserPrivilegeType::Create => "CREATE",
+            UserPrivilegeType::Update => "UPDATE",
             UserPrivilegeType::Select => "SELECT",
             UserPrivilegeType::Insert => "INSERT",
+            UserPrivilegeType::Delete => "DELETE",
+            UserPrivilegeType::Drop => "DROP",
+            UserPrivilegeType::Alter => "ALTER",
+            UserPrivilegeType::Super => "SUPER",
+            UserPrivilegeType::CreateUser => "CREATE USER",
+            UserPrivilegeType::CreateRole => "CREATE ROLE",
+            UserPrivilegeType::Grant => "GRANT",
             UserPrivilegeType::Set => "SET",
         })
     }
 }
 
+/// Each privilege has a set of contexts, which limits the grant objects which the privilege can
+/// take effects on, for example, we can not grant a SUPER privilege (which is Global level)
+/// to a table (in a table level).
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum UserPrivilegeContext {
```

Comment:
Snowflake's permission model has a clear hierarchy:
* Securable object: An entity to which access can be granted. Unless allowed by a grant, access will be denied.

* Role: An entity to which privileges can be granted. Roles are in turn assigned to users. Note that roles can also be assigned to other roles, creating a role hierarchy.

* Privilege: A defined level of access to an object. Multiple distinct privileges may be used to control the granularity of access granted.

* User: A user identity recognized by Snowflake, whether associated with a person or program.

https://docs.snowflake.com/en/user-guide/security-access-control-overview.html
 
Do we have the plan to split the UserPrivilegeContext to:
1. Object: Database, Table, Role, User
2. Roles: SystemAdmin

?

Reply:
Currently we have a GrantObject struct which resembles the Object in the snowflake model, and we have a plan to add the Role.

the UserPrivilegeContext mainly effects on the SHOW PRIVILEGES statement from mysql, it looks like:

https://dev.mysql.com/doc/refman/8.0/en/show-privileges.html

```
mysql> SHOW PRIVILEGES\G
*************************** 1. row ***************************
Privilege: Alter
  Context: Tables
  Comment: To alter the table
*************************** 2. row ***************************
Privilege: Alter routine
  Context: Functions,Procedures
  Comment: To alter or drop stored functions/procedures
*************************** 3. row ***************************
Privilege: Create
  Context: Databases,Tables,Indexes
  Comment: To create new databases and tables
*************************** 4. row ***************************
Privilege: Create routine
  Context: Databases
  Comment: To use CREATE FUNCTION/PROCEDURE
*************************** 5. row ***************************
Privilege: Create temporary tables
  Context: Databases
  Comment: To use CREATE TEMPORARY TABLE
...

```

besides this, the UserPrivilegeContext struct do not really have got many other benefits yet. I think we can skip the context field yet, until we found it nesscary in our privilege model.

here is our current model in planned:

- a GrantEntry has many privileges and works on an Object
- an Object can be Global, a Database or a Table (or Columns in the future)
- an User or Role has a GrantSet, which contains multiple grant entries
- an User can bind with multiple roles, the GrantEntries from the roles which binds to the user + the GrantEntries the user directly got, produces the effective grant set

Reply:
Got, thank you!